include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -43,54 +133,76 @@ struct DefaultKernelName;
 template <typename KernelName = DefaultKernelName>
 class device_policy
 {
+    static sycl::queue
+    __get_default_queue()
+    {
+        static sycl::queue __q(sycl::default_selector_v);
+        return __q;
+    }
+
   public:
     using kernel_name = KernelName;
 
-    device_policy() = default;
+    device_policy() : device_policy(__get_default_queue()) {}
     template <typename OtherName>
-    device_policy(const device_policy<OtherName>& other) : q(other.queue())
-    {
-    }
-    explicit device_policy(sycl::queue q_) : q(q_) {}
-    explicit device_policy(sycl::device d_) : q(d_) {}
-    operator sycl::queue() const { return q; }
+    device_policy(const device_policy<OtherName>& other) : __qh(other.queue()) {}
+    explicit device_policy(sycl::queue q) : __qh(q) {}
+    explicit device_policy(sycl::device d) : __qh(d) {}
+
+    operator sycl::queue() const { return queue(); }
     sycl::queue
     queue() const
     {
-        return q;
+        return __qh.__get_queue();
     }
 
+#if _ONEDPL_PREDEFINED_POLICIES
+    explicit device_policy(__internal::__global_instance_tag __t) : __qh(__t, /*factory*/__get_default_queue) {}
+
+  protected:
+    device_policy(__internal::__global_instance_tag __t, __internal::__queue_factory __f) : __qh(__t, __f) {}
+#endif
+
   private:
-    sycl::queue q;
+    static_assert(sizeof(__internal::__queue_holder) == sizeof(sycl::queue));
+    static_assert(alignof(__internal::__queue_holder) == alignof(sycl::queue));
+    __internal::__queue_holder __qh;
 };
 
 #if _ONEDPL_FPGA_DEVICE
 struct DefaultKernelNameFPGA;
+
 template <unsigned int factor = 1, typename KernelName = DefaultKernelNameFPGA>
 class fpga_policy : public device_policy<KernelName>
 {
     using base = device_policy<KernelName>;
 
+    static sycl::queue
+    __get_fpga_default_queue()
```

Comment:
Probably we may combine this function with `__get_fpga_default_queue()` and pass different arguments as params.
In this case this function `__get_fpga_default_queue()` will not required anymore.

Reply:
No, we cannot; an important part is the static `queue` instance that the function "holds", which needs to be different for the `dpcpp_fpga` policy.

Reply:
Understood, agree. You are right.